testing/integration/upgrade_test.go

```diff
@@ -157,3 +163,170 @@ func getPreviousMinorVersion(v string) (string, error) {
+	// if we want to sort in descending order we could use
+	sort.Sort(sort.Reverse(sortedParsedVersions))
+
+	var latestSnapshotVersion *version.ParsedSemVer
+	// fetch the latest SNAPSHOT build
+	for _, pv := range sortedParsedVersions {
+		if pv.IsSnapshot() {
+			latestSnapshotVersion = pv
+			break
+		}
+	}
+
+	s.Require().NotNil(latestSnapshotVersion)
+
+	// get all the builds of the snapshot version (need to pass x.y.z-SNAPSHOT format)
+	builds, err := aac.GetBuildsForVersion(ctx, latestSnapshotVersion.VersionWithPrerelease())
+	s.Require().NoError(err)
+	// TODO if we don't have at least 2 builds, select the next older snapshot build
+	s.Require().Greater(len(builds.Builds), 1)
+
+	// take the penultimate build of the snapshot (the builds are ordered from most to least recent)
+	upgradeVersionString := builds.Builds[1]
+
+	s.T().Logf("Targeting build %q of version %q", upgradeVersionString, latestSnapshotVersion)
+
+	buildDetails, err := aac.GetBuildDetails(ctx, latestSnapshotVersion.VersionWithPrerelease(), upgradeVersionString)
+	s.Require().NoErrorf(err, "error accessing build details for version %q and buildID %q", latestSnapshotVersion.Original(), upgradeVersionString)
+	s.Require().NotNil(buildDetails)
+	agentBuildDetails, ok := buildDetails.Build.Projects["elastic-agent"]
+	s.Require().Truef(ok, "elastic agent project not found in version %q build %q", latestSnapshotVersion.Original(), upgradeVersionString)
+
+	expectedAgentHashAfterUpgrade := agentBuildDetails.CommitHash
+
+	buildFragments := strings.Split(upgradeVersionString, "-")
+	s.Require().Lenf(buildFragments, 2, "version %q returned by artifact api is not in format <version>-<buildID>", upgradeVersionString)
+
+	upgradeInputVersion := version.NewParsedSemVer(
+		latestSnapshotVersion.Major(),
+		latestSnapshotVersion.Minor(),
+		latestSnapshotVersion.Patch(),
+		latestSnapshotVersion.Prerelease(),
+		buildFragments[1],
+	)
+
+	s.T().Logf("Upgrading to version %q", upgradeInputVersion)
+
+	err = c.Connect(ctx)
+	s.Require().NoError(err, "error connecting client to agent")
+	defer c.Disconnect()
+
+	_, err = c.Upgrade(ctx, upgradeInputVersion.String(), "", false)
+	s.Require().NoErrorf(err, "error triggering agent upgrade to version %q", upgradeInputVersion.String())
+
+	s.Require().Eventuallyf(func() bool {
+		state, err := c.State(ctx)
+		if err != nil {
+			s.T().Logf("error getting the agent state: %v", err)
+			return false
+		}
+		s.T().Logf("current agent state: %+v", state)
+		return state.Info.Commit == expectedAgentHashAfterUpgrade && state.State == cproto.State_HEALTHY
+	}, 10*time.Minute, 1*time.Second, "agent never upgraded to expected version")
+
+	updateMarkerFile := filepath.Join(paths.DefaultBasePath, "Elastic", "Agent", "data", ".update-marker")
+
+	s.Require().FileExists(updateMarkerFile)
+
+	// The checks of the update marker makes the test time out since it runs for more than 10 minutes :(
```

Comment:
Yeah I think we need to figure out if we can shorten this somehow for out testing purposes. Waiting 10 minutes is not really something that any developer wants to wait on.

Reply:
Yeah I just wonder if we should see if we can shorten the time windows somehow just for the test.

Reply:
Probably not for this PR, but we could enhance the watcher to try and look for an optional `watcher.properties` or `watcher.test.properties` file.  This file could hold an override value for the `gracePeriodDuration` that's used by the watcher (as well as any other similar settings used by the error checker and crash checker).  Then, (only) tests could create this file before kicking off the upgrade.

Reply:
Opened https://github.com/elastic/elastic-agent/issues/2796

Reply:
@pchila Could you create an issue to discuss how to shorten the watcher's grace period duration and possibly other similar duration-related settings and link it from the commented out code in this PR?  That way we don't forget about this conversation here.  After that, I'm good to LGTM this PR.

Reply:
@pchila Looks like you created the issue just as I was typing my previous comment 🙂.  Thanks.  Just reference it from the commented out code in the test here and I'lll LGTM this PR.